In [1]:
import torch
print(torch.__version__)

2.11.0+cu128


In [2]:
if torch.cuda.is_available():
    print("GPU is available!")
    print(f"Using GPU:{torch.cuda.get_device_name(0)}")
else:
    print("GPU not Avalilable: Using CPU")

GPU is available!
Using GPU:NVIDIA GeForce RTX 2050


In [3]:
torch.cuda?

Type:        module
String form: <module 'torch.cuda' from 'c:\\Users\\DSingh\\miniconda3\\envs\\ecg_cs\\Lib\\site-packages\\torch\\cuda\\__init__.py'>
File:        c:\users\dsingh\miniconda3\envs\ecg_cs\lib\site-packages\torch\cuda\__init__.py
Docstring:  
This package adds support for CUDA tensor types.

It implements the same function as CPU tensors, but they utilize
GPUs for computation.

It is lazily initialized, so you can always import it, and use
:func:`is_available()` to determine if your system supports CUDA.

:ref:`cuda-semantics` has more details about working with CUDA.

In [4]:
torch.cuda.get_device_name?

Signature: torch.cuda.get_device_name(device: torch.device | str | int | None = None) -> str
Docstring:
Get the name of a device.

Args:
    device (torch.device or int or str, optional): device for which to return the
        name. This function is a no-op if this argument is a negative
        integer. It uses the current device, given by :func:`~torch.cuda.current_device`,
        if :attr:`device` is ``None`` (default).

Returns:
    str: the name of the device
File:      c:\users\dsingh\miniconda3\envs\ecg_cs\lib\site-packages\torch\cuda\__init__.py
Type:      function

## Creating Tensors

In [6]:
#empty tensor   it creates the specified size of the tensor and not initialize it (it just provide the memory) 
a= torch.empty(2,3)
print(a)
print(type(a))

tensor([[-9.9814e+27,  1.8231e-42,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00]])
<class 'torch.Tensor'>


In [8]:
#Using Zeroes    it create the specified size of the tensor and initialize with zero
b=torch.zeros(2,3)
print(b)
print(type(b))

tensor([[0., 0., 0.],
        [0., 0., 0.]])
<class 'torch.Tensor'>


In [9]:
#Using Ones   it create the specified size of the tensor and initialize with 1
b=torch.ones(2,3)
print(b)
print(type(b))

tensor([[1., 1., 1.],
        [1., 1., 1.]])
<class 'torch.Tensor'>


In [10]:
#Using rand   it create the specified size of the tensor and intilialize with the random value between 0 and 1

c=torch.rand(2,3)
print(c)
print(type(c))

tensor([[0.7969, 0.1347, 0.7421],
        [0.2879, 0.3793, 0.8538]])
<class 'torch.Tensor'>


In [11]:
torch.manual_seed(13)
torch.rand(2,3)

tensor([[0.0918, 0.4794, 0.8106],
        [0.0151, 0.0153, 0.6036]])

In [12]:
torch.manual_seed(13)
torch.rand(2,3)

tensor([[0.0918, 0.4794, 0.8106],
        [0.0151, 0.0153, 0.6036]])

In [13]:
# Using tensor   it create the custom tensor 
torch.tensor([[1,2,3],[4,5,6]])

tensor([[1, 2, 3],
        [4, 5, 6]])

In [ ]:
#other ways to creates the tensor 
#sometimes it will be useful the use them also

#arange
print('using arange ->',torch.arange(0,10))

#linespace
print('using linspace ->',torch.linspace(0,10,10))

#eye    eye for identity matrix
print('using eye ->',torch.eye(10))

#using full
print('using full ->',torch.full((3,3),10))

using arange -> tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
using linspace -> tensor([ 0.0000,  1.1111,  2.2222,  3.3333,  4.4444,  5.5556,  6.6667,  7.7778,
         8.8889, 10.0000])
using eye -> tensor([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])
using full -> tensor([[10, 10, 10],
        [10, 10, 10],
        [10, 10, 10]])


## Tensor Shape

In [126]:
x=torch.tensor([[1,2,3,4],[5,6,7,8]])
x

tensor([[1, 2, 3, 4],
        [5, 6, 7, 8]])

In [19]:
x.shape

torch.Size([2, 4])

In [22]:
## when we want to create the tensor of the same shape 
torch.empty_like(x)

tensor([[5591745954176,             0,             0,             0],
        [            0,             0,             0,             0]])

In [23]:
torch.zeros_like(x)

tensor([[0, 0, 0, 0],
        [0, 0, 0, 0]])

In [24]:
torch.ones_like(x)


tensor([[1, 1, 1, 1],
        [1, 1, 1, 1]])

In [127]:
## error comes due to the datatype missmatched between the x(contains all the integer value) and 
## rand generates the float between [0,1]
torch.rand_like(x)    

NotImplementedError: "check_uniform_bounds" not implemented for 'Long'

## Tensor Data Types

In [26]:
# using dtype

x.dtype

torch.int64

In [27]:
# assign data types

torch.tensor([1.0,2.0,3.5], dtype=torch.int32)

tensor([1, 2, 3], dtype=torch.int32)

In [28]:
torch.tensor([1,2,3], dtype=torch.float64)


tensor([1., 2., 3.], dtype=torch.float64)

In [29]:
## using to      to change the datatypes of the current tensor 
x.to(torch.float64)

tensor([[1., 2., 3., 4.],
        [5., 6., 7., 8.]], dtype=torch.float64)

In [30]:
## now using torch.rand_like() with the datatype conversion
print(x.dtype)
print(torch.rand_like(x,dtype=torch.float32))

torch.int64
tensor([[0.2318, 0.8633, 0.9859, 0.1975],
        [0.0830, 0.4253, 0.9149, 0.4799]])


## datatype table copied from the CampusX google colab notebook of the tensor in pytorch for the reference 

| **Data Type**             | **Dtype**         | **Description**                                                                                                                                                                |
|---------------------------|-------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **32-bit Floating Point** | `torch.float32`   | Standard floating-point type used for most deep learning tasks. Provides a balance between precision and memory usage.                                                         |
| **64-bit Floating Point** | `torch.float64`   | Double-precision floating point. Useful for high-precision numerical tasks but uses more memory.                                                                               |
| **16-bit Floating Point** | `torch.float16`   | Half-precision floating point. Commonly used in mixed-precision training to reduce memory and computational overhead on modern GPUs.                                            |
| **BFloat16**              | `torch.bfloat16`  | Brain floating-point format with reduced precision compared to `float16`. Used in mixed-precision training, especially on TPUs.                                                |
| **8-bit Floating Point**  | `torch.float8`    | Ultra-low-precision floating point. Used for experimental applications and extreme memory-constrained environments (less common).                                               |
| **8-bit Integer**         | `torch.int8`      | 8-bit signed integer. Used for quantized models to save memory and computation in inference.                                                                                   |
| **16-bit Integer**        | `torch.int16`     | 16-bit signed integer. Useful for special numerical tasks requiring intermediate precision.                                                                                    |
| **32-bit Integer**        | `torch.int32`     | Standard signed integer type. Commonly used for indexing and general-purpose numerical tasks.                                                                                  |
| **64-bit Integer**        | `torch.int64`     | Long integer type. Often used for large indexing arrays or for tasks involving large numbers.                                                                                  |
| **8-bit Unsigned Integer**| `torch.uint8`     | 8-bit unsigned integer. Commonly used for image data (e.g., pixel values between 0 and 255).                                                                                    |
| **Boolean**               | `torch.bool`      | Boolean type, stores `True` or `False` values. Often used for masks in logical operations.                                                                                      |
| **Complex 64**            | `torch.complex64` | Complex number type with 32-bit real and 32-bit imaginary parts. Used for scientific and signal processing tasks.                                                               |
| **Complex 128**           | `torch.complex128`| Complex number type with 64-bit real and 64-bit imaginary parts. Offers higher precision but uses more memory.                                                                 |
| **Quantized Integer**     | `torch.qint8`     | Quantized signed 8-bit integer. Used in quantized models for efficient inference.                                                                                              |
| **Quantized Unsigned Integer** | `torch.quint8` | Quantized unsigned 8-bit integer. Often used for quantized tensors in image-related tasks.                                                                                     |


## Mathematical Operation on the Tensor

## 1. Scaler Operation

In [31]:
x=torch.rand(2,2)

In [32]:
x

tensor([[0.5348, 0.2695],
        [0.2530, 0.3390]])

In [34]:
#addition
print("Addition-> ", x+2)
#subtraction
print("Subtraction-> ", x-2)
#multiplication
print("Multiplication-> ", x*3)
#division
print("Division-> ", x/3)
#integer division
print("integer division-> ", (x*100)//3)
#mod
print("mod-> ", ((x*100)//3)%2)
#power
print("Power-> ", x**2)

Addition->  tensor([[2.5348, 2.2695],
        [2.2530, 2.3390]])
Subtraction->  tensor([[-1.4652, -1.7305],
        [-1.7470, -1.6610]])
Multiplication->  tensor([[1.6044, 0.8084],
        [0.7591, 1.0170]])
Division->  tensor([[0.1783, 0.0898],
        [0.0843, 0.1130]])
integer division->  tensor([[17.,  8.],
        [ 8., 11.]])
mod->  tensor([[1., 0.],
        [0., 1.]])
Power->  tensor([[0.2860, 0.0726],
        [0.0640, 0.1149]])


## 2. Element Wise Operation

In [35]:
a=torch.rand(2,3)
b=torch.rand(2,3)

In [36]:
print(a)
print(b)

tensor([[0.8367, 0.1289, 0.9693],
        [0.4495, 0.4031, 0.8202]])
tensor([[0.9792, 0.3278, 0.4076],
        [0.7276, 0.4506, 0.2334]])


In [37]:
#addition
print("Addition-> ", a+b)
#subtraction
print("Subtraction-> ", a-b)
#multiplication
print("Multiplication-> ", a*b)
#division
print("Division-> ", a/b)
#mod
print("mod-> ", a%b)
#power
print("Power-> ", a**b)

Addition->  tensor([[1.8159, 0.4567, 1.3770],
        [1.1771, 0.8537, 1.0536]])
Subtraction->  tensor([[-0.1425, -0.1989,  0.5617],
        [-0.2781, -0.0476,  0.5869]])
Multiplication->  tensor([[0.8193, 0.0422, 0.3951],
        [0.3271, 0.1816, 0.1914]])
Division->  tensor([[0.8544, 0.3932, 2.3782],
        [0.6178, 0.8944, 3.5148]])
mod->  tensor([[0.8367, 0.1289, 0.1541],
        [0.4495, 0.4031, 0.1201]])
Power->  tensor([[0.8398, 0.5109, 0.9874],
        [0.5589, 0.6640, 0.9548]])


In [39]:
## using abs for finding the absolute value

c= torch.tensor([1,-2,3,-4])
print(c)
torch.abs(c)

tensor([ 1, -2,  3, -4])


tensor([1, 2, 3, 4])

In [40]:
## using negative  for negate the value inside the tensor
torch.negative(c)

tensor([-1,  2, -3,  4])

In [41]:
## using round   for round off the digits in the tensor

d=torch.tensor([1.9, 1.5, 3.7, 2.3, 4.5])
print(d)
print(torch.round(d))

tensor([1.9000, 1.5000, 3.7000, 2.3000, 4.5000])
tensor([2., 2., 4., 2., 4.])


In [42]:
## we can now perform the ceil and floor also on the tensor

print(torch.ceil(d))
print(torch.floor(d))


tensor([2., 2., 4., 3., 5.])
tensor([1., 1., 3., 2., 4.])


In [43]:
## using clamp    it will fix the number in range and
##  the number in out of the range become the boundry value

torch.clamp(d, min=2, max=3)

tensor([2.0000, 2.0000, 3.0000, 2.3000, 3.0000])

## 3. Reduction Operation

In [49]:
e=torch.randint(size=(2,4), low=0, high=10, dtype=torch.float32)

In [50]:
print(e)

tensor([[4., 8., 0., 2.],
        [8., 5., 3., 5.]])


In [51]:
## sum all the element in tensor
print(torch.sum(e))

## sum along columns
print(torch.sum(e, dim=0))

## sum along rows
print(torch.sum(e, dim=1))

tensor(35.)
tensor([12., 13.,  3.,  7.])
tensor([14., 21.])


In [52]:
## mean also works as the sum 
## mean all the element in tensor
print(torch.mean(e))

## sum along columns
print(torch.mean(e, dim=0))

## sum along rows
print(torch.mean(e, dim=1))


tensor(4.3750)
tensor([6.0000, 6.5000, 1.5000, 3.5000])
tensor([3.5000, 5.2500])


In [53]:
## similarly median and max and min

## median
print(torch.median(e))

## max
print(torch.min(e))

## max
print(torch.max(e))

tensor(4.)
tensor(0.)
tensor(8.)


In [56]:
## product all the element in the tensor
print(torch.prod(e))

## standard deviation of the tensor
print(torch.std(e))


## variance of the tensor
print(torch.var(e))

tensor(0.)
tensor(2.7742)
tensor(7.6964)


In [58]:
## using argmax   it will show the position of the largest element in the tensor

print(torch.argmax(e))
print(torch.argmin(e))

tensor(1)
tensor(2)


## 4. matrix Operation

In [59]:
f=torch.randint(size=(2,3), low=1, high=10)
g=torch.randint(size=(3,4), low=1, high=10)

print(f)
print(g)

tensor([[5, 2, 6],
        [8, 2, 9]])
tensor([[8, 3, 3, 4],
        [2, 6, 2, 8],
        [5, 4, 6, 9]])


In [60]:
## matrix multiplication
print(torch.matmul(f,g))

tensor([[ 74,  51,  55,  90],
        [113,  72,  82, 129]])


In [ ]:
vector1 = torch.tensor([1,2])
vector2 = torch.tensor([3,4])

# dot product
torch.dot(vector1, vector2)   ## 1*3 + 2*4 == 11

tensor(11)

In [65]:
## transpose of the matrix
print(f)
print(torch.transpose(f, dim0=0 , dim1=1))

tensor([[5, 2, 6],
        [8, 2, 9]])
tensor([[5, 8],
        [2, 2],
        [6, 9]])


In [70]:
h=torch.randint(size=(3,3), low=0, high=10, dtype=torch.float32)
print(h)

tensor([[2., 4., 6.],
        [5., 0., 9.],
        [6., 6., 0.]])


In [71]:
print(torch.det(h))

tensor(288.)


In [72]:
# inverse
torch.inverse(h)

tensor([[-0.1875,  0.1250,  0.1250],
        [ 0.1875, -0.1250,  0.0417],
        [ 0.1042,  0.0417, -0.0694]])

## 5. Comparison Operations 

In [73]:

i=torch.randint(size=(2,3), low=0, high=10)
j=torch.randint(size=(2,3), low=0, high=10)

print(i)
print(j)

tensor([[5, 9, 6],
        [3, 1, 3]])
tensor([[7, 5, 6],
        [5, 6, 9]])


In [74]:
## now all the comparison operation I'll use here

## greater than
print(i>j)

## less than
print(i<j)

## equal to
print(i==j)

## not equal to
print(i!=j)

## greater than equal to
print(i>=j)

## less than equal to
print(i<=j)

tensor([[False,  True, False],
        [False, False, False]])
tensor([[ True, False, False],
        [ True,  True,  True]])
tensor([[False, False,  True],
        [False, False, False]])
tensor([[ True,  True, False],
        [ True,  True,  True]])
tensor([[False,  True,  True],
        [False, False, False]])
tensor([[ True, False,  True],
        [ True,  True,  True]])


## 6. special Functions

In [75]:
k=torch.randint(size=(2,3), low=0, high=10)

print(k)

tensor([[4, 1, 2],
        [1, 7, 1]])


In [76]:
## log
torch.log(k)

tensor([[1.3863, 0.0000, 0.6931],
        [0.0000, 1.9459, 0.0000]])

In [77]:
## exp
torch.exp(k)

tensor([[  54.5981,    2.7183,    7.3891],
        [   2.7183, 1096.6332,    2.7183]])

In [78]:
## sqrt

torch.sqrt(k)

tensor([[2.0000, 1.0000, 1.4142],
        [1.0000, 2.6458, 1.0000]])

In [79]:
## sigmoid
torch.sigmoid(k)

tensor([[0.9820, 0.7311, 0.8808],
        [0.7311, 0.9991, 0.7311]])

In [80]:
## softmax
kk=k.to(dtype=torch.float32)
torch.softmax(kk, dim=0)

tensor([[0.9526, 0.0025, 0.7311],
        [0.0474, 0.9975, 0.2689]])

In [81]:
torch.relu(k)

tensor([[4, 1, 2],
        [1, 7, 1]])

# Inplace Operations

In [82]:
m=torch.rand(2,3)
n=torch.rand(2,3)
print(m,n,sep="\n")

tensor([[0.6378, 0.0647, 0.3546],
        [0.1348, 0.3244, 0.2068]])
tensor([[0.2808, 0.8788, 0.8631],
        [0.8353, 0.8574, 0.1544]])


In [83]:
## the addition operation create the new tensor 
m+n

tensor([[0.9186, 0.9434, 1.2177],
        [0.9701, 1.1818, 0.3611]])

### sometimes we require the changes in the original tensor so that the memory will not occupy more 

to do this we require the function and apply the underscore(_) after the function \
example: relu(m)  ---> m.relu_() \
         m+n  ----> m.add_(n) \
         softmax(m) ----> m.softmax_()

In [ ]:
m.add_(n)  ## this is treated as the inplace operations

tensor([[0.9186, 0.9434, 1.2177],
        [0.9701, 1.1818, 0.3611]])

In [85]:
m

tensor([[0.9186, 0.9434, 1.2177],
        [0.9701, 1.1818, 0.3611]])

In [86]:
### similary we can apply it on other operation also

m.relu_()

tensor([[0.9186, 0.9434, 1.2177],
        [0.9701, 1.1818, 0.3611]])

In [87]:
m

tensor([[0.9186, 0.9434, 1.2177],
        [0.9701, 1.1818, 0.3611]])

## Copying a Tensor

In [88]:
a=torch.rand(2,3)
a

tensor([[0.0546, 0.4962, 0.1069],
        [0.9401, 0.8407, 0.6069]])

#### generally for copying we use the assignment operator but it has some problem....
#### it will create the reference of the tensor variable

In [89]:
b=a
print(b)
a[0][0]=0
print(a)
print(b)   ## now b will also change 

tensor([[0.0546, 0.4962, 0.1069],
        [0.9401, 0.8407, 0.6069]])
tensor([[0.0000, 0.4962, 0.1069],
        [0.9401, 0.8407, 0.6069]])
tensor([[0.0000, 0.4962, 0.1069],
        [0.9401, 0.8407, 0.6069]])


In [90]:
## to prove this we can use the id() function

print(id(a))
print(id(b))

2298757716304
2298757716304


In [91]:
## for better copying we can use the clone function......

b=a.clone()

In [92]:
print(id(a))
print(id(b))

2298757716304
2298758345056


## tensor operation on GPU

In [93]:
device = torch.device('cuda')

In [94]:
## creating a new tensor on GPU
torch.rand((2,3), device=device)

tensor([[0.4524, 0.9270, 0.8315],
        [0.5458, 0.7006, 0.7442]], device='cuda:0')

In [100]:
## moving an existing tensor to GPU
a=torch.rand(2,3)
print(a)
b = a.to(device)
print(b)

tensor([[0.0065, 0.6250, 0.3426],
        [0.0959, 0.9077, 0.7391]])
tensor([[0.0065, 0.6250, 0.3426],
        [0.0959, 0.9077, 0.7391]], device='cuda:0')


In [101]:
## this operation is performed on the GPU
b+5

tensor([[5.0065, 5.6250, 5.3426],
        [5.0959, 5.9077, 5.7391]], device='cuda:0')

### Comparison on the basis of time between the tensor running on the CPU and the GPU

In [103]:
import time

#Define the size of the matrices
size=10000 # Large size for performance comparison

# Create the random matrices on CPU
matrix_cpu1=torch.randn(size,size)
matrix_cpu2=torch.randn(size,size)

## measure time on CPU
start_time=time.time()
result_cpu=torch.matmul(matrix_cpu1, matrix_cpu2)
cpu_time=time.time()-start_time

print(f"Time on CPU: {cpu_time:.4f} seconds")

# Move matrix to GPU
matrix_gpu1=matrix_cpu1.to('cuda')
matrix_gpu2=matrix_cpu2.to('cuda')

## measure time on GPU
start_time=time.time()
result_gpu=torch.matmul(matrix_gpu1, matrix_gpu2)
torch.cuda.synchronize()  ## ensure all the GPU operation are complete
gpu_time=time.time()-start_time

print(f"Time on GPU: {gpu_time:.4f} seconds")

## compare results
print("\n Speedup (CPU time / GPU time): ", cpu_time/gpu_time)

Time on CPU: 4.9571 seconds
Time on GPU: 1.5042 seconds

 Speedup (CPU time / GPU time):  3.295420792510763


## Reshaping the Tensor

In [106]:
a=torch.ones(4,4)
a

tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])

In [107]:
## reshape
a.reshape(2,2,2,2)

tensor([[[[1., 1.],
          [1., 1.]],

         [[1., 1.],
          [1., 1.]]],


        [[[1., 1.],
          [1., 1.]],

         [[1., 1.],
          [1., 1.]]]])

In [ ]:
 ## flatten
 a.flatten()

tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [109]:
b=torch.rand(2,3,4)
b

tensor([[[0.8503, 0.0180, 0.4307, 0.8720],
         [0.5471, 0.7483, 0.4701, 0.0951],
         [0.4612, 0.4120, 0.4823, 0.6792]],

        [[0.4164, 0.3023, 0.7343, 0.3864],
         [0.5663, 0.0869, 0.5307, 0.7430],
         [0.7433, 0.4316, 0.2980, 0.6500]]])

In [111]:
## permute
## here order 2,1,0 representing the order of the shape of the b tensor
## orginal shape(2,3,4) having order is 0,1,2 
# we change it to shape(4,3,2) order become 2,1,0 
print(b.permute(2,1,0))
print(b.permute(2,1,0).shape)


tensor([[[0.8503, 0.4164],
         [0.5471, 0.5663],
         [0.4612, 0.7433]],

        [[0.0180, 0.3023],
         [0.7483, 0.0869],
         [0.4120, 0.4316]],

        [[0.4307, 0.7343],
         [0.4701, 0.5307],
         [0.4823, 0.2980]],

        [[0.8720, 0.3864],
         [0.0951, 0.7430],
         [0.6792, 0.6500]]])
torch.Size([4, 3, 2])


above let's take an example \
in original---> b[0][0][1]= 0.0180
in permute ----> b[1][0][0] = 0.0180

In [113]:
## unsqueeze
#image size
c=torch.rand(226,226,3)
print(c.unsqueeze(0).shape)
print(c.unsqueeze(1).shape)
print(c.unsqueeze(2).shape)
print(c.unsqueeze(3).shape)


torch.Size([1, 226, 226, 3])
torch.Size([226, 1, 226, 3])
torch.Size([226, 226, 1, 3])
torch.Size([226, 226, 3, 1])


In [116]:
d=torch.rand(1,20)
## shape[1,20]

d.squeeze(0).shape

torch.Size([20])

## Numpy and Pytorch

### we can convert the tensor of the pytorch into numpy array and vice versa

In [117]:
import numpy as np
a=torch.tensor([1,2,3])
a

tensor([1, 2, 3])

In [120]:
b=a.numpy()
b

array([1, 2, 3])

In [121]:
type(b)

numpy.ndarray

In [123]:
c=np.array([1,2,3])
c

array([1, 2, 3])

In [124]:
torch.from_numpy(c)

tensor([1, 2, 3])